web_infrastructure/ansible_tower.py

```diff
@@ -0,0 +1,343 @@
+'''
+
+RETURN = '''
+job_result:
+    description: The Ansible Tower job result dictionary
+    returned: always
+    type: dict
+'''
+
+try:
+    import requests
+    from requests.packages.urllib3.exceptions import InsecureRequestWarning
+    # Suppress insecure warnings if using this module asynchronously
+    requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
+    HAS_REQUESTS = True
+except ImportError:
+    HAS_REQUESTS = False
+
+from ansible.module_utils.basic import AnsibleModule
+from time import sleep
+import json, sys
+
+# API paths
+AUTH_PATH = "/api/v1/authtoken/"
+JOB_PATH = "/api/v1/jobs/"
+JOB_TEMPLATE_PATH = "/api/v1/job_templates/"
+POLL_INTERVAL = 10
+
+class AnsibleTowerManager:
+    """Handles Ansible Tower Jobs"""
+    def __init__(self, params):
+        self.host = params['host']
+        self.user = params['user']
+        self.password = params['password']
+        self.verify_ssl = params['verify_ssl']
+        self.extra_vars = params['extra_vars']
+        self.cloud_credential = params['cloud_credential']
+        self.credential = params['credential']
+        self.inventory = params['inventory']
+        self.tags = params['tags']
+        self.limit = params['limit']
+        self.timeout = params['timeout']
+        self.session = requests.Session()
+        self.session.headers.update({'Content-Type':'application/json'})
+        
+        # Authenticate
+        self.authenticate()
+        
+    def authenticate(self):
+        auth_body = dict(username=self.user, password=self.password)
+        auth_response = self.session.post(url=self.host + AUTH_PATH, data=json.dumps(auth_body), verify=self.verify_ssl)
+        auth_response.raise_for_status()
+        auth_token = json.loads(auth_response.text).get('token')
+        self.session.headers.update({'Authorization':'Token ' + auth_token})
+
+    def poll_job(self, job_id):
+        poll_count = self.timeout / POLL_INTERVAL + 1
+        counter = 0
+        job_status = 'pending'
+        job_url = self.host + JOB_PATH + str(job_id) + "/"
+        while counter < poll_count and job_status in ['pending', 'waiting', 'running']:
+            sleep(POLL_INTERVAL)
+            job_response = self.session.get(url=job_url, verify=self.verify_ssl)
+            job_response.raise_for_status()
+            job_result = json.loads(job_response.text)
+            job_status = job_result.get('status')
+            counter += 1
+        return job_result
+
+    def launch_job(self, job_template_id, launch_data):
+        launch_url = self.host + JOB_TEMPLATE_PATH + str(job_template_id) + "/launch/"
+        launch_job_response = self.session.post(url=launch_url, data=json.dumps(launch_data), verify=self.verify_ssl)
+        launch_job_response.raise_for_status()
+        launch_job_data = json.loads(launch_job_response.text)
+        job_id = launch_job_data.get('id') or launch_job_data.get('job')
+        return self.poll_job(job_id)
```

Comment:
The way that tower-cli allows for synchronous running is that it takes it accepts a "monitor" parameter, which turns polling on or off. I would think that's the type of approach that you would want in this module, and your code is already modularized in a way which should make that easy.

I'm still unsure if it makes the most sense to import tower-cli or not. I think there could be some arguments to avoid it. The approach I would personally prefer would be to see what problems we would have integrating it with the work you've started here (like setting the token as a separate method), and see what work on the tower-cli side would be required to make that flow smoothly.

We also have a `verify_ssl` parameter in tower-cli, and people have noted that it can be pretty useless without the ability to point it to a local certificate. Development for that is still in-process for tower-cli.